core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/FileSink.java

```diff
@@ -0,0 +1,263 @@
+import co.cask.hydrator.common.ReferencePluginConfig;
+import co.cask.hydrator.common.batch.JobUtils;
+import co.cask.hydrator.common.batch.sink.SinkOutputFormatProvider;
+import co.cask.hydrator.plugin.common.StructuredToAvroTransformer;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.avro.mapreduce.AvroJob;
+import org.apache.avro.mapreduce.AvroKeyOutputFormat;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
+import org.apache.parquet.avro.AvroParquetOutputFormat;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * Writes to the FileSystem.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name("File")
+@Description("Writes to the FileSystem.")
+public class FileSink extends ReferenceBatchSink<StructuredRecord, Object, Object> {
+  private static final String NULL_STRING = "\0";
+  private Conf config;
+  private Function<StructuredRecord, KeyValue<Object, Object>> recordTransformer;
+
+  public FileSink(Conf config) {
+    super(config);
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate();
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws IOException {
+    // if user provided macro, need to still validate
+    config.validate();
+    Job job = JobUtils.createInstance();
+    Configuration hConf = job.getConfiguration();
+    hConf.set(FileOutputFormat.OUTDIR, config.getOutputDir(context.getLogicalStartTime()));
+    for (Map.Entry<String, String> entry : config.getFSProperties().entrySet()) {
+      hConf.set(entry.getKey(), entry.getValue());
+    }
+
+    Schema schema = config.getSchema();
+    if (schema == null) {
+      schema = context.getInputSchema();
+    }
+
+    if ("text".equals(config.format)) {
+      context.addOutput(Output.of(config.referenceName,
+                                  new SinkOutputFormatProvider(TextOutputFormat.class, hConf)));
+    } else if ("avro".equals(config.format)) {
+      if (schema == null) {
+        throw new IllegalStateException("Cannot write using avro format when schema is unknown. "
+                                          + "Please set the schema explicitly.");
```

Comment:
should these checks also be part of validate during config ? when its non-macro.

Reply:
true, will do